src/platform_storage_api/admission_controller/volume_resolver.py

```diff
@@ -0,0 +1,219 @@
+import dataclasses
+import logging
+from enum import Enum
+from pathlib import PurePath
+from types import TracebackType
+from typing import Any, Optional
+
+from apolo_kube_client.client import KubeClient
+
+from platform_storage_api.config import AdmissionControllerConfig
+from platform_storage_api.storage import StoragePathResolver
+
+
+logger = logging.getLogger(__name__)
+
+
+class VolumeResolverError(Exception):
+    ...
+
+
+class VolumeBackend(str, Enum):
+    """Supported volume backends"""
+    NFS = "nfs"
+
+
+@dataclasses.dataclass
+class BaseVolumeSpec:
+
+    def to_kube(self) -> dict[str, Any]:
+        return dataclasses.asdict(self)
+
+
+@dataclasses.dataclass
+class NfsVolumeSpec(BaseVolumeSpec):
+    server: str
+    path: str
+
+    @classmethod
+    def from_pv(cls, pv: dict[str, Any]) -> "NfsVolumeSpec":
+        """
+        Constructs a volume spec from a PV object
+        """
+        return cls(
+            server=pv["spec"][VolumeBackend.NFS]["server"],
+            path=pv["spec"][VolumeBackend.NFS]["path"]
+        )
+
+
+T_VolumeSpec = NfsVolumeSpec  # we'll add others later?
+
+
+@dataclasses.dataclass
+class KubeVolume:
+    backend: VolumeBackend
+    spec: T_VolumeSpec
+
+    def to_kube(self) -> dict[str, Any]:
+        return {
+            self.backend.value: self.spec.to_kube()
+        }
+
+
+class KubeVolumeResolver:
+
+    def __init__(
+        self,
+        kube_client: KubeClient,
+        path_resolver: StoragePathResolver,
+        admission_controller_config: AdmissionControllerConfig,
+    ):
+        self._kube = kube_client
+        self._path_resolver = path_resolver
+        self._config = admission_controller_config
+
+        self._local_fs_prefix_to_kube_volume: dict[str, KubeVolume] = {}
+        """will store a mapping, where a key is a local mounted path,
+        and a value is a kube volume definition.
+        """
+
+    async def __aenter__(self) -> "KubeVolumeResolver":
+        """
+        Initialize the kube volume resolver.
+        Gets the most fresh platform-storage POD and passes it to an
+        internal state updater, so we can resolve the real mount paths.
+        """
+        logger.info("initializing volume resolver")
+        namespace_url = self._kube.generate_namespace_url()
+
+        # get all storage admission controllers PODs,
+        # and choose the freshest one
+        pods_response = await self._kube.get(
+            f"{namespace_url}/pods",
+            params={
+                "labelSelector": f"app={self._config.service_name}",
+            }
+        )
```

Comment:
I think it's better to get the current pod

Reply:
@zubenkoivan right! fixed this. used `socked.gethostname()`